packages/SwingSet/src/liveslots/virtualObjectManager.js

```diff
@@ -585,13 +593,47 @@ export function makeVirtualObjectManager(
   ) {
     const {
       finish,
+      stateShape = undefined,
       thisfulMethods = false,
       interfaceGuard = undefined,
     } = options;
     let facetNames;
     let contextMapTemplate;
     let prototypeTemplate;
 
+    harden(stateShape);
+    stateShape === undefined ||
+      passStyleOf(stateShape) === 'copyRecord' ||
+      assert.fail(X`A stateShape must be a copyRecord: ${q(stateShape)}`);
+    assertPattern(stateShape);
+
+    const serializeSlot = (slotState, prop) => {
+      if (stateShape !== undefined) {
+        hasOwnPropertyOf(stateShape, prop) ||
+          assert.fail(
+            X`State must only have fields described by stateShape: ${q(
+              ownKeys(stateShape),
+            )}`,
+          );
+        fit(slotState, stateShape[prop], prop);
+      }
+      return serialize(slotState);
+    };
+
+    const unserializeSlot = (slotData, prop) => {
+      const slotValue = unserialize(slotData);
+      if (stateShape !== undefined) {
+        hasOwnPropertyOf(stateShape, prop) ||
+          assert.fail(
+            X`State only has fields described by stateShape: ${q(
+              ownKeys(stateShape),
+            )}`,
+          );
+        fit(slotValue, stateShape[prop]);
+      }
```

Comment:
Given that we check on store, we probably should not redundantly check on read.

Is there any similar issue when reading data stored by a previous incarnation?

Reply:
Resolving.